Implemented performance improvement for TriviaHelper.GetContainingTriviaList

vweijsters opened the pull request:
Fixes #957 

Tested it using the StyleCopTester pull request.
In release mode, running SA1512 over the Microsoft/CodeContracts solution resulted in:

Scenario | Result (time spent in TriviaHelper.GetContainingTriviaList)
----- | -----
Original | 12764 ms (average)
New | 1962 ms (average)

Big thanks to @pdelvo for helping out (= doing a lot of the work) with this one.

pdelvo commented on StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TriviaHelper.cs:
```diff
@@ -14,20 +17,20 @@ internal static class TriviaHelper
-                case SyntaxKind.EndOfLineTrivia:
-                case SyntaxKind.WhitespaceTrivia:
-                    break;
+                    case SyntaxKind.EndOfLineTrivia:
```
You should probably revert thise indentation changes

vweijsters replied:
Will do. The auto formatter of Visual Studio got me again :disappointed:

pdelvo commented on StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1512SingleLineCommentsMustNotBeFollowedByBlankLine.cs:
```diff
@@ -170,9 +171,9 @@ private static bool IsOnOwnLine(SyntaxTriviaList triviaList, int triviaIndex)
+        private static bool IsPartOfFileHeader(IReadOnlyList<SyntaxTrivia> triviaList, int triviaIndex)
         {
-            if (triviaList.FullSpan.Start > 0)
+            if (triviaList[0].FullSpan.Start > 0)
```
Can you be sure that triviaList is not empty?

vweijsters replied:
No, probably not.

pdelvo commented:
Thats a really significant improvement. Using the StyleCopTester with roslyn 1.0.0 running all our analyzers on Microsoft/CodeContracts gave me these results (green arrows are improvements)

![comparison](https://cloud.githubusercontent.com/assets/1710209/8687263/d9df75a4-2a8e-11e5-8c9f-4de14e2d9407.PNG)

pdelvo commented on StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TriviaHelper.cs:
```diff
@@ -130,19 +133,30 @@ internal static SyntaxTriviaList WithoutLeadingWhitespace(this SyntaxTriviaList
+            else
+            {
+                var prevToken = token.GetPreviousToken();
+                triviaIndex = prevToken.TrailingTrivia.Count + token.LeadingTrivia.IndexOf(trivia);
```
triviaIndex is not used. This is actually a (relativly) big performance hit. Thats how I noticed it.

sharwell replied:
:question: Is this something that still needs to be fixed before the change is merged?

pdelvo replied:
You could remove this when merging. This shows up prominently in performance reports

vweijsters replied:
:exclamation: You cannot remote that. `triviaIndex` is an out parameter and several callers of the method require that value.

pdelvo replied:
You're right. I overlooked that

vweijsters commented:
I also removed all instances of SyntaxTriviaList.AddRange. 
This will address some of the issues in #964.

I added it in the PR because it needed the helper class introduced here.
